Groups

erget opened the pull request:
Dear community,

This is PR fixes https://github.com/cf-convention/cf-conventions/issues/144. Charlie and I look forward to your input and hope that we can finish integrating this into the Convention soon, as discussed at the CF meeting on 20 June 2018.

Notable changes: The proposed changes to the Convention would regulate the use of groups in CF files. How variables are resolved across group boundaries is described, and a new chapter of the Appendix provides examples of how these ideas are currently used by several data providers, as well as how these would match the CF Conventions.

~~See http://cf-trac.llnl.gov/trac/ticket/XXX~~ - Pursuant discussion in https://github.com/cf-convention/cf-conventions/issues/130 no Trac ticket has been created.

 - [x] Added link from ~~trac ticket~~ **Github issue** to this PR?

I'm not sure what this means, so I'm leaving this blank:

 > Applied via ​https://github.com/cf-convention/cf-conventions/pull/XXX

 - [x] Updated "Revision History"? (Use the date you applied the changes.)

davidhassell commented on ch02.adoc:
```diff
@@ -163,5 +163,66 @@ The global **`external_variables`** attribute is a blank-separated list of the n
+
+* `title`
+* `history`
+* `conventions`
```
```suggestion
* `Conventions`
```

davidhassell commented on appi.adoc:
```diff
@@ -0,0 +1,397 @@
+
+. The use of attributes attached to a group in order to store metadata normally attached directly to variables is discouraged.
+This includes, for example, replacing per-variable attributes like `_FillValue`, `scale_factor`, `valid_min`, with group-level equivalents.
+Although group attributes might be more concise, it is likely to create problems with downstream software and to reduce interoperability.
```
This actually sounds dangerous! It is very likely that different variables in the group would have different `scale_factor` values, even assuming they had all been scaled. There are a number of named attributes that do not describe the physical nature of variables, but rather describe the encoding of a data array. These (which include the three that you have named) shouldn't these be specifically prohibited from being set on the group?  What do you think?

czender replied:
@davidhassell 
The recommendation _discourages_ using a group-level `scale_factor` to prevent the situation you mention. Perhaps you misread it or I do not understand your concern. Or are you saying the practice should be "forbidden", not merely discouraged?

erget replied:
Left as-is for the purpose of minimising changes at this stage.

davidhassell commented on appi.adoc:
```diff
@@ -0,0 +1,397 @@
+For example, storage of ensembles as sibling groups (as described below) is often clearer when the realization number is encoded in the group name.
+In this case, the `Realization` group attribute retains the realization number even if the group is renamed.
+Other commonly enumerated group names, such as station identifiers, buoy numbers, or channel wavelengths are analogous.
+It is fine to enumerate or itemize names so long as the number or name is redundantly stored as a group attribute.
```
I would _much_ rather say that a group attribute is the canonical way forward, but that you can redundantly use enumerated or itemized group names.

davidhassell replied:
What I meant to say is, I think: I would rather just drop the word "redundant" - the group attribute is metadata according to convention, after all.

czender replied:
@davidhassell I agree. "redundant" is not the best wording. I suggest we replace it with "also".

erget replied:
I'll leave it as-is unless someone has strong feelings for the purpose of minimising changes at this stage.

erget replied:
Oops, sorry for the noise - I thought I was commenting elsewhere. I've adopted the wording in 9e5fd72.

davidhassell commented on appi.adoc:
```diff
@@ -0,0 +1,397 @@
+When the coordinates are outside the group that contains the `coordinates` attribute, it is tempting to store the coordinate locations as full, unambiguous paths such as `"/g1/lat /g1/lon"`, or as relative paths such as `"g1/lat g1/lon"`.
+However, paths that contain slashes must be explicitly altered when the variable is subset into a new file with a different group hierarchy, or when the hierarchy is flattened.
+CF attributes affected by this practice include `ancillary_variables`, `bounds`, `cell_measures`, `climatology`,  `coordinates`, `formula_terms`, and `grid_mapping`.
+The simpler `"lat lon"` specification works without alteration in all situations for out-of-group locations so long as it is understood to mean the nearest identifiers that have the referring variable in their scope.
```
Just a question: is `coordinates="/g1/lat /g1/lon"` _ever_ a good idea, in practice? You're argument that the search rules should be instead in preference is good, and I can't think of any exceptions, off hand. Perhaps there are legacy datasets that need catering for?

czender replied:
@davidhassell Your comment is slightly garbled. Many NASA legacy datasets currently use the `"/g1/lat /g1/lon"` paradigm. The lateral search algorithm would allow them to maintain the same group structure while eliminating the absolute paths.

davidhassell replied:
OK - no problem, apologies for not making myself clearer. I consider the question answered, and the text is fine.

davidhassell commented on appi.adoc:
```diff
@@ -0,0 +1,397 @@
+The simpler `"lat lon"` specification works without alteration in all situations for out-of-group locations so long as it is understood to mean the nearest identifiers that have the referring variable in their scope.
+
+. When using a grid mapping which is referenced in multiple groups, the reference to the coordinates themselves should be maintained as an attribute of the referencing variables, and not as an attribute of the grid mapping itself. This way the coordinates can resolve to the correct coordinates which are "owned" by the referencing variable, and which are not shared amongst variables, while the gri ...
++
```
I don't really get this, I'm afraid. CF doesn't let you store coordinate variables on a grid mapping variable, anyway, so what is the need for this advice, here?

erget replied:
For me the advantage here is extra clarity. There are some cases where the same grid mapping is re-used, but the coordinates are different, for example if you're re-using an orthographic coordinate system in a netCDF file with radiances from different instruments viewing with different resolutions but on the same platform - basically two variables might share `grid_mapping_name` but have different variables storing `projection_x_coordinate` and `projection_x_coordinate`.

Our teams found this clearer but perhaps this is just over-specifying. If this principal is clear without this clause I'm happy to drop the clause.

davidhassell replied:
OK, I see where you're coming from. I think you're right that it is worth highlighting that groups don't change the standard behaviour here. Perhaps this minor change would make it clear that there is nothing surprising here? 

"When using a grid mapping which is referenced in multiple groups, the reference to the
coordinates themselves should be maintained as an attribute of the referencing variables,
**in the usual manner (ref to section 5.6)**. This way the coordinates can resolve to the correct
coordinates which are "owned" by the referencing variable, and which are not shared amongst
variables, while the grid mapping which is used to interpret the coordinates is stored in only one place. Both coordinates and grid mapping are thus found via regular scoping mechanisms."

czender commented:
replying to @martinjuckes above
> Hello, I support the intention and approach here and don't want to restart the clock as far as approval goes, but I have a question about inheritance. In the 2nd paragraph of recommended practises in Appendix A, the text states that the group should be self contained, taking into account inherited content. My understanding is that the NetCDF4 interpretation of inheritance is such that a group only inherits from the top level, is that correct and what is intended here? 

I do know where you get that impression of netCDF4 inheritance. The netCDF4 implementation of the CDM has groups inheriting from their immediate ancestors and all the way up to root, in that order of precedence, and that is what this proposal recommends and intends.

> The 3rd paragraph, however, describes a much looser approach to inheriting information based on proximity, so that "lat", for example, should be interpreted in terms of the nearest object with that name. This sounds fragile to me, and appears to contradict the recommendation for self-contained object. 

Per my preceding response, I think you will agree that the inheritance model proposed does not contradict the recommendation for self-contained objects.

> Could the recommendation be amended to specify that variables referenced in this way are recommended to have unique names within the file?

It could be, but this would exclude important use cases as discussed in #144. Furthermore "variables referenced in this way" includes coordinates located by direct ancestor search as well as lateral search. Direct ancestor search has not been controversial thus far, and lateral search is discouraged by this note in Ch. 01 line 210: 
```
[NOTE]
====
This use of the lateral search strategy to find them is discouraged.
They are allowed mainly for backwards-compatibility with existing datasets, and may be deprecated in future versions of the standard.
====
```

czender commented on ch02.adoc:
```diff
@@ -163,5 +163,76 @@ The global **`external_variables`** attribute is a blank-separated list of the n
+
+A variable or dimension specified with no path (for example, `lat`) refers to the variable or dimension of that name, if there is one, in the referring group.
+If not, the ancestors of the referring group are searched for it, starting from the direct ancestor and proceeding toward the root group, until it is found.
+
```
Please give an example so we can be sure to address your concern, which is unclear to me.

czender replied:
Pinging @marqh to please give an example, preferably in CDL, of the situation about which you think a note should be added.

marqh replied:
Hello @czender 

I'd like to reuse the examples that I think should be included in this PR, but are currently missing

do you know if thiis interpretation is correct, is there a file missing from this PR, which includes example CDL with groups?

thank you
mark

czender replied:
I am unsure whether Appendix I will be in the initial PR. Please read the following exchange between @JonathanGregory and @erget (from #144):

>> Could we please omit Appendix I for the moment from the pull request? By the way, it is entitled Appendix A.

> I don't really understand this request, to be honest, because I think that it would make it easier to interpret the additions, but I've done it for the moment ([fcad2b7](https://github.com/cf-convention/cf-conventions/commit/fcad2b7cddc264b2ccb7bf38876d0a3b4f5a5257)). Also, the reason it's called "Appendix A" in the rendering is because the appendix numbering actually takes place during the compilation of the Conventions as a whole, so when you view the individual file, the `[appendix]` tag is converted into "Appendix A" due to the lack of context. If you compile the complete Conventions it's titled appropriately.
> 
> I don't think that the appendix is really a break from convention, especially in light of Appendix H, which is large and comprehensive, as well as extremely useful, and would prefer to keep it, but I'll defer to your judgement.

marqh replied:
I also think that the appendix adds significant informational value and should be included.

Not withstanding that opinion, I think that this PR should be self consistent. That suggests to me either:
* re-include the appendix
* remove references to the appendix from this PR

My preference is to re-include the appendix.

If @JonathanGregory has particular reasons why not including the appendix is preferred, then I'd be interested to hear them. @JonathanGregory please may you share your view with us here?  thank you.

If non-inclusion is the prevailing view, then I think it preferable that the references are removed so that a merged PR is self-consistent.

erget replied:
@JonathanGregory would you be willing to weigh in on this? As stated in #144 I do agree with @marq here that it would be better to include the appendix. My opinion is also that any issues that we have with an appendix can be addressed more easily should we find them, as the appendix is consistent with the proposed changes to chapter 2, so the only issues should be editorial.

@marqh in the meantime I'm removing the references in order to keep with the discussion in #144, but this is isolated to 664e10f so we can revert the commit later if needed. As stated in previous discussions, I am *in favour* of keeping Appendix I and agree with your assertion that it adds informational value.

marqh replied:
thank you for following this through, 

the change, for consistency, is sensible, but the preference from me would be to revert this change and reintroduce the appendix, as I have previously stated.

however, this change is sufficient to remove my objection

martinjuckes replied:
Hello, I'm sorry if this has been dealt with elsewhere, but it looks to me as though the [NASA EOS file format convention](https://earthdata.nasa.gov/user-resources/standards-and-references/hdf-eos5) could break the proposed lateral search algorithm. As such EO datasets have been advanced as a reason for including support for implicit references outside the direct parental line of a group, this would be a problem. 
It may be that I have misunderstood the term "local apex group": is this defined anywhere?

czender replied:
@martinjuckes The proposed CF lateral search algorithm has evolved over time from a stronger (and more complex and fragile) form that did support most NASA EOS datasets to a weaker (and less complex and fragile) form that will not support many NASA EOS datasets. Perhaps the most relevant change is that the lateral search algorithm now proposed is restricted to single-dimension NUG-style coordinates, rather than the multidimensional auxiliary coordinates originally proposed. Please see #144 for more discussion. I would characterize this as the proposed CF convention not supporting NASA EOS rather than NASA EOS breaking the proposed convention. The NASA EOS convention predates the proposed CF Groups convention by many years. 
This PR defines the term "local apex group" is defined in the ch01.adoc glossary as

>  "local apex group:: The nearest (to a referring group) ancestor group in which a dimension underlying an out-of-group coordinate is defined. The word "apex" refers to position of this group at the vertex of the tree of groups formed by it, the referring group, and the group where a coordinate is located."

czender commented on ch02.adoc:
```diff
@@ -163,5 +163,76 @@ The global **`external_variables`** attribute is a blank-separated list of the n
+For example, a `coordinates` attribute of `g1/lat` refers to the `lat` variable in subgroup `g1` of the current (referring) group.
+Upward path traversals from the current group are indicated with the UNIX convention.
+For example, `../g1/lat` refers to the `lat` variable in the sibling group `g1` of the current (referring) group.
+
```
When a variable references something, let's say a coordinate, that includes an upward traversal in the hierarchical tree, then upward traversal via relative paths become useful. Many (most?) variables only reference out-of-group (OOG) objects that include an upward traversal. This is because the most commonly referenced OOG objects share dimensions with the referencing variable, and dimensions must be defined at or above the level of the referencing variable. By contrast, "downward traversal" has many fewer applications. So eliminating upward but not downward traversal would drastically curtail the use of relative path searches. Perhaps the two strongest arguments for relative vs. absolute path searches are that 1.Relative paths are less fragile that absolute because entire branches of a hierarchy can be moved between datasets without altering any of the relative paths. 2. Relative paths are shorter and thus produce more legible metadata.  As for an example, consider an ensemble of temperature data from different models all regridded to the same grid. The coordinates would naturally be stored in the parent group, and the child groups would naturally each contain a temperature variable referencing the coordinates in the parent group.

marqh replied:
Hello @czender 

thank you for following up on this comment. I perceive that the benefit of upward traversal is valuable.

given the section:

```
A variable or dimension specified with no path (for example, `lat`) refers to the variable or dimension of that name, if there is one, in the referring group.
If not, the ancestors of the referring group are searched for it, starting from the direct ancestor and proceeding toward the root group, until it is found.
```
Do you think there is value in having both of these mechanisms enabling upward traversal, and thus having 2 different mechanisms to support?

Would it be easier to manage if only one of these were provided?

Are there already cases of both relative paths with `../` and of references with no path requiring ancestor searching in wide use, such that both mechanisms must be supported, or is one much more common and could be targeted as the supported mechanism?

thank you
mark

czender replied:
My short answer is yes. There is value in having three search methods (ancestor, relative, absolute). IMHO ancestor is the most robust and logical choice to design a dataset with. Relative search is fine and less likely to break as branches are pruned into other files. Absolute paths are more precise yet fragile than the other two. YMMV.

Regarding current practice, most hierarchical netCDF4 datasets that I have seen are from satellite retrievals, and those that attempt to be CF-compliant use absolute paths. I have not seen relative paths. I think I have seen ancestor search, yet I cannot recall which datasets that was with.

The effort required to support absolute paths is the least, since those are valid input to the netCDF API. Relative paths require marginally more effort than absolute paths. Some pathname manipulation is required before accessing the netCDF API. The effort required to support ancestor search is putting a loop around the search. Neither of those algorithms is difficult to implement. The cost of supporting relative and ancestor search must of course be weighed agains the benefits of increased flexibility and robustness.

marqh replied:
Hello @czender 

this makes a lot of sense, thank you

I have found an explicit reference within the NetCDF documentation regarding scope for dimensions for groups:

```
Dimensions are scoped such that they are visible to all child groups. For example, you can define a dimension in the root group, and use its dimension id when defining a variable in a sub-group.
```
https://www.unidata.ucar.edu/software/netcdf/docs/groups.html

I wonder if it is worth repeating and referencing this within the CF documentation?  It was not immediately clear to me whether the scoping visibility of parent dimensions within child groups was part of this CF proposal, or part of netCDF (which it is).

(there may also be a better reference for this than the one I provide)

This could give improved context for the scope visibility presented here.

erget replied:
I have included this text in f3f4675. Does this seem clearer now?

marqh replied:
that's helpful, thank you

erget commented on history.adoc:
```diff
@@ -1,6 +1,9 @@
 
+.3 August 2018
+. Added <<groups, section on using groups>> and <<appendix-examples-groups,examples of good practices>>.
+
```
You're right about this. I'm used to working with revision histories that are sorted with the newest changes at the top (I think that's the most common case), so I instinctually inserted these changes at the top. I've moved it down in 9649b0e. Thanks for catching that.

erget commented on toc-extra.adoc:
```diff
@@ -75,3 +75,8 @@ H.19. <<example-h.19>>
+I.2. <<example-i.2>>
+I.3. <<example-i.3>>
+I.4. <<example-i.4>>
+I.5. <<example-i.5>>
```
Removed in 9649b0e, they're a part of Appendix I (insert sad face).

erget commented on cf-conventions.adoc:
```diff
@@ -94,6 +94,9 @@ include::appg.adoc[]
 
+:numbered:
+include::appi.adoc[]
+
```
Removed in 664e10f.

erget commented:
Dear @marqh - thanks very much for your productive comments and my apologies for not responding sooner - first I was [building robots with students in Africa](http://fullstackembedded.com) and since then I've been in the hospital for malaria ;) Actually I still am, but I'm happy to have this diversion.

Concerning the "missing" appendix - discussions in #144 with @JonathanGregory have resulted in me removing it. I personally feel that this is a shame, as the other appendices are very helpful and in my opinion this appendix would be helpful in the same way. @JonathanGregory feels it is the best way forward in order to accelerate the adoption of the changes to Chapter 2, since the new appendix should be carefully reviewed - fair enough, and I'm willing to follow this two-step approach to adoption, even if it's not my preferred approach. Unfortunately, when I removed the appendix, I forgot to remove the references to the appendix that you note - I've done that in the subsequent commits based on your feedback.

I believe I've addressed all of your review items in the commits above - please let me know how you find the changes.

dblodgett-usgs commented:
@erget -- I believe this has all been agreed to, correct? It has been long enough to go ahead and merge if you can get the conflicts fixed up?

erget commented:
@dblodgett-usgs done.

erget commented:
@dblodgett-usgs do we have a checklist for items that have to be updated here? These haven't been done / captured yet:
- Update to author list (didn't do it for this PR, didn't think of it)
- Update to conformance doc (https://github.com/cf-convention/Conformance/pull/8)

dblodgett-usgs commented:
We should add a checklist to the template for a PR. The template above is what we have right now. Want to add the author list update?

Is the conformance up to date and ready to merge too?

erget commented:
I've opened #180 with the author change. IMHO the conformance doc update in cf-convention/Conformance#8 is OK but it hasn't been reviewed; I've pinged Jonathan to ask him if he'd be willing to, as he moderated #144.

erget commented:
@RosalynHatcher thanks, please be aware that the proposed changes to the Conformance document are in cf-convention/Conformance#8
